File: BT/PaintingHouses/PaintingHouses.java
/*
[5, 3, 4],
[2, 1, 6],
[2, 3, 4],
[4, 3, 3]

Note: this is not the most optimal solution.
*/
import java.util.*;

class Solution {
    public int solve(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return 0;
        int[][] dp = new int[matrix.length][matrix[0].length];
        for (int row = 0; row < dp.length; row++) Arrays.fill(dp[row], -1);
        return solve(matrix, 0, -1, dp);
    }

    private int solve(int[][] matrix, int row, int prevCol, int[][]dp) {
        if (row == matrix.length) return 0;
        if (prevCol > -1 && dp[row][prevCol] != -1) return dp[row][prevCol];

        int lowest = Integer.MAX_VALUE;
        for (int col = 0; col < matrix[0].length; col++) {
            if (prevCol == col) continue;
            int cost = matrix[row][col];
            lowest = Math.min(lowest, cost + solve(matrix, row + 1, col, dp));
        }

        if (prevCol > -1) {
            dp[row][prevCol] = lowest;
        }
        return lowest;
    }
}

File: BT/PaintingHouses/PaintingHouses.py
class Solution:
    def solve(self, matrix):
        if not matrix or not matrix[0]:
            return 0
        
        m, n = len(matrix), len(matrix[0])
        
        for row in range(m - 2, -1, -1):
            next_list = matrix[row+1][:]
            m1, m2 = self.getTwoMins(next_list)

            for col in range(n):
                matrix[row][col] += m1[1] if m1[0] != col else m2[1]
        
        return min(matrix[0])
    
    def getTwoMins(self, l):
        m1, m2 = None, None

        for i, val in enumerate(l):
            if m1 is None:
                m1 = (i, val)
                continue

            if m2 is None or val < m2[1]:
                m2 = (i, val)
            
            if m2[1] < m1[1]:
                m1, m2 = m2, m1

        return m1, m2
